server/src/main/java/org/elasticsearch/env/NodeMetadata.java

```diff
@@ -162,15 +163,15 @@ public void setOldestIndexVersion(int oldestIndexVersion) {
             this.oldestIndexVersion = IndexVersion.fromId(oldestIndexVersion);
         }
 
+        @UpdateBeforeV9 // version is required in the node metadata from v9 onwards
+        private Version getVersionOrFallbackToEmpty() {
+            return Objects.requireNonNullElse(this.nodeVersion, Version.V_EMPTY);
+        }
+
         public NodeMetadata build() {
-            final Version nodeVersion;
+            final Version nodeVersion = getVersionOrFallbackToEmpty();
```

Comment:
Can put the annotation on the variable directly here

Reply:
This case is borderline I think, both works, but it's super nice in the other case you pointed out because it allows to skip the introduction of a method just to annotate it. Thanks!